### PR TITLE
Incorrectly using a non-generic type with type parameters now produces a helpful Python error instead of throwing NullReferenceException

### LINKED ISSUE #1325: NullReferenceException when non-generic type is given type parameters
### Environment

-   Pythonnet version: [latest git commit](https://github.com/pythonnet/pythonnet/commit/bd264cf003ca1f9727d7eaeb90ddf913e81fdad2)
-   Python version: 3.7
-   Operating System: Windows 10 64-bit

### Details

The following incorrect Python code produces a confusing NullReferenceException, instead of a more helpful error message:
```python
from System.Collections import IList
IList[bool]
```

-   If there was a crash, please include the traceback here.

```
Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
   at Python.Runtime.GenericUtil.GenericByName(String ns, String name, Int32 paramCount)
   at Python.Runtime.ClassBase.type_subscript(IntPtr idx)
```

A more helpful message would be: `System.Collections.IList is not a generic type`

### PR DESCRIPTION
**Pull request by tminka**
### What does this implement/fix? Explain your changes.

Fixes #1325 

### Does this close any currently open issues?

Fixes #1325

### Any other comments?

...

### Checklist

Check all those that are applicable and complete.

-   [X] Make sure to include one or more tests for your change
-   [ ] If an enhancement PR, please create docs and at best an example
-   [X] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [X] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/pythonnet/pythonnet?pullRequest=1326) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/pythonnet/pythonnet?pullRequest=1326) before we can accept your contribution.<br/><br/>:x: tminka [sign now](https://cla.dotnetfoundation.org/pythonnet/pythonnet?pullRequest=1326)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/pythonnet/pythonnet?pullRequest=1326) it.</sub>

**Comment by tminka** [comment]
The tests appear to be failing for the reasons addressed by PR #1260.  Applying those changes should make the tests pass.

**Comment by lostmsu** [comment]
@tminka I am a bit lost. In the comment to that PR you mentioned, that https://github.com/pythonnet/pythonnet/commit/de7e1ac776545d5834f5d9cc85480d60d8604860 solves the issue. But the PR itself does not contain that commit.

File: src/runtime/genericutil.cs [hunk 1]
```diff
@@ -75,20 +75,27 @@ public static List<string> GetGenericBaseNames(string ns)
...
             return GenericByName(t.Namespace, t.Name, paramCount);
         }
 
+        /// <summary>
+        /// Finds a generic type in the given namespace with the given name and number of generic parameters.
+        /// </summary>
         public static Type GenericByName(string ns, string name, int paramCount)
         {
-            foreach (Type t in GenericsByName(ns, name))
+            var types = GenericsByName(ns, name);
```
Comment (filmor) [inline src/runtime/genericutil.cs:90]: Is this function used somewhere else? Maybe it should just not return `null`.

**Review by filmor — CHANGES_REQUESTED** [review]
Very nice addition, thank you. If you can just move the test to Python, this is good to go :)

File: src/runtime/genericutil.cs (hunk 1 above)
Comment (tminka) [reply src/runtime/genericutil.cs:90]: It is also used by GenericsForType, but that function seems to be unused itself.  Another option is to return an empty list.  However, that would involve an allocation.  Array.Empty makes more sense, which would involve some more changes.

File: src/embed_tests/TestList.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        }
+
+        [Test]
+        public void MissingGenericTypeTest()
+        {
+            Assert.Throws<PythonException>(() =>
+            PythonEngine.Exec($@"
+from System.Collections import IList
+IList[bool]
+"));
```
[the remark this answers is not in the document]
Comment (tminka) [reply src/embed_tests/TestList.cs:32]: There are a few problems with Python tests at the moment:
1. tests.pyproj is not part of pythonnet.sln, and has the old configurations still in it.
1. Python tests run under the current Python environment, which requires installing the development version of Python.NET into the current Python environment.  Anyone who forgets to do this will be testing the wrong version of Python.NET.
1. If Python.Runtime throws an exception (as in this case), the Visual Studio debugger will not be activated.
Using an embedded test avoids all of these problems.

File: src/runtime/genericutil.cs (hunk 1 above)
Comment (tminka) [reply src/runtime/genericutil.cs:90]: Perhaps the best approach is to inline GenericsByName into GenericByName.

File: src/embed_tests/TestList.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lostmsu) [reply src/embed_tests/TestList.cs:32]: @tminka here's how I deal with this:
1. Open python tests code in VS Code in a separate window, or add tests.pyproj to .sln
2. `conda create -n pynet-dev-py39 python=3.9`
3. Open `Python.Runtime` project settings in VS, and create a new profile like this:
image

But true, these problems cause me to often prefer .NET tests over Python tests.

Maybe we should come up with a better way to run these tests from Visual Studio.

**Comment by tminka** [comment]
@lostmsu My only point was that the test failures are the same as the ones in the current master, and those are due to finalization/deallocation bugs.

File: src/embed_tests/TestList.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tminka) [reply src/embed_tests/TestList.cs:32]: 1. Adding tests.pyproj to sln doesn't work right because of the old configurations in there.  It also changes the sln file, which will show up in the commit.
2. Creating a virtual environment doesn't really solve the problem.  You still need to update Python.NET within that environment.  Referring to that environment also changes tests.pyproj which changes the commit.
3. This runs but doesn't stop at any breakpoints or exceptions.  It won't stop even if you raise an error in Python.

Basically this is no replacement for the simplicity of cloning the repository, opening the sln in Visual Studio, building, and running tests there.

[the remark this answers is not in the document]
Comment (tminka) [reply src/embed_tests/TestList.cs]: I have found a solution that should make everyone happy.  The new test is a Python test, and there is a runner that can execute any Python test as an embedded test.

[the remark this answers is not in the document]
Comment (lostmsu) [reply src/embed_tests/TestList.cs]: Let's keep the runner in a separate PR. Meanwhile the test failures in CI have been fixed.

**Comment by filmor** [comment]
Please make this just a normal Python test. You can run pytest from within VisualStudio using our normal APIs (`import pytest; pytest.main()`) if you want.

**Comment by filmor** [comment]
Ah, I see that you have already moved the test. Would you be okay with splitting the `EmbeddedPythonTest` into a separate PR? I think this does have merit, but it's independent of the actual fix here.
